routes/merged/project/location/load_balancer.rb

```diff
@@ -0,0 +1,158 @@
+        end
+      end
+
+      r.post "detach-vm" do
+        Authorization.authorize(current_account.id, "LoadBalancer:edit", lb.id)
+        required_parameters = %w[vm_id]
+        request_body_params = Validation.validate_request_body(json_params, required_parameters)
+
+        vm = Vm.from_ubid(request_body_params["vm_id"])
+        unless vm
+          response.status = 404
+
+          if api?
+            r.halt
+          else
+            flash["error"] = "VM not found"
+            r.redirect "#{@project.path}#{lb.path}"
+          end
+        end
+
+        Authorization.authorize(current_account.id, "Vm:view", vm.id)
+
+        lb.detach_vm(vm)
+
+        if api?
+          Serializers::LoadBalancer.serialize(lb, {detailed: true})
+        else
+          flash["notice"] = "VM is detached"
+          r.redirect "#{@project.path}#{lb.path}"
+        end
+      end
+
+      r.get true do
+        Authorization.authorize(current_account.id, "LoadBalancer:view", lb.id)
+        @lb = Serializers::LoadBalancer.serialize(lb, {detailed: true, vms_serialized: !api?})
+        if api?
+          @lb
+        else
+          vms = lb.private_subnet.vms_dataset.authorized(current_account.id, "Vm:view").exclude(Sequel[:vm][:id] => lb.vms_dataset.select(Sequel[:vm][:id])).all
+          @attachable_vms = Serializers::Vm.serialize(vms)
+
+          view "networking/load_balancer/show"
+        end
+      end
+
+      r.delete true do
+        Authorization.authorize(current_account.id, "LoadBalancer:delete", lb.id)
+        lb.incr_destroy
+        response.status = 204
+        r.halt
+      end
+
+      r.patch api? do
+        Authorization.authorize(current_account.id, "LoadBalancer:edit", lb.id)
+        request_body_params = Validation.validate_request_body(json_params, %w[algorithm src_port dst_port health_check_endpoint vms])
+        lb.update(
+          algorithm: request_body_params["algorithm"],
+          src_port: Validation.validate_port(:src_port, request_body_params["src_port"]),
+          dst_port: Validation.validate_port(:dst_port, request_body_params["dst_port"]),
+          health_check_endpoint: request_body_params["health_check_endpoint"]
+        )
+
+        request_body_params["vms"].each do |vm_id|
+          vm_id = vm_id.delete("\"")
+          vm = Vm.from_ubid(vm_id)
+          unless vm
+            response.status = 404
+            r.halt
+          end
+
+          Authorization.authorize(current_account.id, "Vm:view", vm.id)
+          if vm.load_balancer && vm.load_balancer.id != lb.id
+            fail Validation::ValidationFailed.new("vm_id" => "VM is already attached to a load balancer")
+          elsif vm.load_balancer && vm.load_balancer.id == lb.id
+            next
+          end
+
+          lb.add_vm(vm)
+        end
+
+        lb.vms.map { _1.ubid }.reject { request_body_params["vms"].map { |vm_id| vm_id.delete("\"") }.include?(_1) }.each do |vm_id|
+          vm = Vm.from_ubid(vm_id)
```

Comment:
Not an issue with this PR, since this was a problem with the previous code, but this section is an N+1 query issue.  Removing the `map { _1.ubid }` and updating the `reject` block would allow removing the `vm = Vm.from_ubid(vm_id)` inside the loop.

Reply:
The routes aren't run in the database transactions, right? Should we put the patch route in to a transaction? If anything goes wrong, we could end up with a partially updated resource.

Reply:
Looks good!

Reply:
Correct.  Anytime we are doing multiple queries in a route, we should surround it in a transaction.  That can be done after the merge, though.